### PR TITLE
Start of Typescript typing file

### PR DESCRIPTION
**Pull request by WorldMaker**
This is the start of a typescript typing file. I'm not sure if this is something you would be interesting keeping in the main repository or not. I think it would be useful for myself and fellow Typescript developers (Typescript now supports lookup of typing information from `node_modules`). It also can be useful for non-Typescript developers. (Editors like Visual Studio Code and Atom can use Typescript definitions to provide better IDE support for JavaScript as well.)

Reasons it is not yet complete/a work in progress:

- More `any` typed things that I would prefer
- Just supports the root import and not sub-imports (`import * as db from 'camo/db'` for instance)
- Doesn't yet have every class or option
- Could use JSDoc comments (ie, copy them from the rest of the project)
- Maybe want Typescript test file too (DefinitelyTyped requires them now on every type definition file these days, which is why I'm contributing less to that monster repo these days)

Anyway, I think this is interesting/useful enough even in this basic state for others to maybe utilize so I'm submitting it as a PR if are interesting in merging it into the repo.

### CODE REVIEW

**Comment by scottwrobinson** [comment]
Hi Max,

Thanks for the PR! I'd like to eventually add more support for Typescript, so this is a good start. I'm thinking I'll start a branch for it initially, and then merge in to master after I've had enough time to evaluate and figure things out.

I'm hoping to have some time this weekend to play around with this and merge.

Thanks again.

Scott

**Comment by jsdream** [comment]
Hello there @scottwrobinson!

+1 for this. Any updates?

**Comment by michaeljota** [comment]
Maybe it's best if you pull this to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped). Idk much about TS, but I like to use those definitions on VS Code to autocomplation and prevent call a no existing funtion.

**Comment by WorldMaker** [comment]
I considered that, Michael, but this point I fear DefinitelyTyped is too big for its britches. It's become too much of a pain to propose new additions to DT. I've been joking about starting an IndefinitivelyTyped alternative, but haven't actually built such a thing. Easier to get library authors to support their own typing.

**Comment by michaeljota** [comment]
I see. Well it is easy to use, but I guess it should be easy to maintain also.

**Comment by WorldMaker** [comment]
Yes, which is the problem with DT in a nutshell: Now that it is so easy to use, they expect new definitions to be nearly perfect and, as the old saying goes, perfect is the enemy of the good. The difference between DT when it started and today is a tale of stark contrasts.

**Comment by michaeljota** [comment]
What about [Typings](https://github.com/typings/typings)?
I think this might solve the problem with TSD. (I think it is being developed by the same people)

**Comment by WorldMaker** [comment]
Neat. I had not seen Typings yet, but I'm excited to learn more and give it a shot.

**Comment by WorldMaker** [comment]
Okay, I started a [typings](https://github.com/typings/typings) targeted repository:

https://github.com/IndefinitivelyTyped/camo

You can already install it with typings by:

    typings install github:IndefinitivelyTyped/camo

I haven't tested it yet, but it should be somewhat more complete than the version in this branch. It also follows the module layout of camo somewhat better. @scottwrobinson is still welcome to consider merging those files here into the main repository (in which case you could `typings install npm:camo`), and/or bless it for the typings registry (`typings install camo`). (Not that it needs an official blessing for the typings registry, I'd just feel more confident that way.)

**Comment by michaeljota** [comment]
Thanks! Will test it and give some fb. :)

**Comment by scottwrobinson** [comment]
Hi all,

Sorry for having ignored this issue for so long. Good to see you're able to make some progress on the typings at least.

To be honest, although I understand TypeScript's goal and main features, I haven't used it enough to understand the best practices and other tools surrounding it (like DefinitelyTyped, Typings, etc), so I'm hesitant to provide a type file out of the box when there might be a better solution.

One of my tasks for the next few weeks is to get this figured out as I'd like Camo to work well with all of the popular transpilers out there, including TypeScript.

Suggestions and feedback are welcome. Thanks!

**Comment by WorldMaker** [comment]
Here's my (longtime Typescript user) view of things, Scott:

Typescript 1.7 added "node module resolution" and with that the modern assumption is that the best path forward for NPM packages is that they bundle their own typing information. The reasons for that should be obvious: if the package maintainer bundles it, it means that the typing information should be reasonably up-to-date and applicable to the version of the package that you install. It also means that in some cases the typing information is auto-generated directly from the source (for instance, packages that themselves are written in Typescript can very easily compile their own typing information).

Typings is a tool built to automate all those cases where a package maintainer doesn't want to include typing information, or cannot for whatever reason. It's goal is essentially to be a typing information package system, trying to connect things like NPM packages to appropriate typing information. But there's still some expectation that in those cases where an NPM package already provides its own typing information, typings should get out of the way for the most part.

DefinitelyTyped is an (outmoded) centralized repository of the typing information for a large number of well known libraries, and prior to "node module resolution" and especially prior to typings was essentially the go to source of typing files. Typings takes a much less centralized, much more package-management style approach (similar to NPM itself, JSPM, Bower, et al), including using DT as a fallback source when that is still useful.

TL;DR: The "best practice" is now for NPM packages to include their own typing information, but only so long as that also reflects the package maintainer's desire to also maintain the typing information. (Given that many JS IDEs are now also using Typescript definitions for autocomplete and such, it's an increasingly good idea to do that, even for projects that don't necessarily want to otherwise make use of Typescript.) That said, Typings is already a fantastic and growing next-best-thing.

Hope those explanations help.

**Comment by WorldMaker** [comment]
Oh, and I'm going to go ahead and close this PR as the [camo typings](https://github.com/IndefinitivelyTyped/camo) is ahead of and more complete than this branch.

**Comment by scottwrobinson** [comment]
@WorldMaker thanks for all the great info and for creating the typing repo. Even with the Typings support, I'll still consider adding typing information to the package.

And for anyone that's interested, @lucasmciruzzi submitted a [PR to DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/8049) for Camo, which was merged today, so Camo is now included!

Thanks everyone!
